### ISSUE TITLE
Merge public-v2 changes into develop?

### ISSUE THREAD

**Issue by edwardhartnett** [NOAA-EMC/NCEPLIBS#151]
@climbfuji I still don't understand the versioning. That's what I get for skipping the last two UFS release meetings! ;-)

We have releases NCEPLIBS-1.0 and NCEPLIBS-1.1 out. And you want to release 2.0? 

You have said:

> UFS MRW Apps 1.0 and 1.1 where released with NCEPLIBS-v1.0.0 (internally ufs-v1.0.0) and NCEPLIBS-v1.1.0 (internally ufs-v1.1.0). The decision was made to ship the SRW App 1.0 with NCEPLIBS-v2.0.0 (internally ufs-v2.0.0).
> 
> The numbering of these release versions does not coincide with the internal number in the develop branch. For each major release of the UFS (same or different App), the public release version of NCEPLIBS will be bumped up if sufficient changes are made. v1.0.0 was an early version of the umbrella build (basically just after we started working on it), and this version here is quite different (pretty close to the development version at the moment).
> 

OK, but why then do we have a separate development version? Shall we not just bump our actual release (in the develop branch) to 2.0 as well? Is there any reason we don't do a full release of NCEPLIBS? Seems like that would be what should be expected.

In the 2.0 release, will all the internal libraries stay at the same version as in 1.1 release?

**Comment by edwardhartnett** [comment]
@climbfuji also when would we like to release 2.0? Once your current PR is merged? Or are there other necessary changes?

**Comment by climbfuji** [comment]
Your nceplibs 1.0 and 1.1 refer to develop. These are a different versioning stream, entirely and unfortunately.

release/public-v1 NCEPLIBS-v1.0.0 and NCEPLIBS-v1.1.0 correspond to nothing in develop, because this was in the very early stages of the umbrella build development, way before the develop 1.0.0 was ever created

release/public-v2 NCEPLIBS-v2.0.0 was cut off develop at the end of June/beginning of July and has since seen updates specifically for the release that will not be sent back to develop (for example this entire PR)

So, there are two parallel versioning streams, one for the public release code, and one for develop. Internally, we refer to the public release versions as ufs-v1.0.0, ufs-v1.1.0, ufs-v2.0.0 to indicate that they are from release/public-vN (N=major version number in ufs-vN.M.R).

I know, this is confusing.

**Comment by climbfuji** [comment]
This is hopefully the last change. Code freeze end of October, followed by one month of testing where only bugfixes are allowed to go in. Release in early December.

**Comment by edwardhartnett** [comment]
Why are the changes in the PR not being merged back to the develop branch? Will they not be needed in the next public release?

**Comment by climbfuji** [comment]
That's a long story ... we make specific adaptations for the release that the developers responsible for the "develop" branches don't want (and don't need).

Also, chances are high that the next public release will be utilizing hpc-stack instead of NCEPLIBS-external and NCEPLIBS.

**Comment by edwardhartnett** [comment]
The public release of NCEPLIBS and NCEPLIBS-external needs to continue to be a cmake release. Going back to custom build scripts like hpc-stack is not an option.

The hpc-stack system is for internal NOAA use, to make sure our HPC systems get set up correctly. It is not for public use. It is not, and should not, be the way we attempt to distribute this software to the user community outside of NOAA.

Though hpc-stack may be of use for other HPC centers, most builders of these packages outside NOAA will do so on linux workstations, not HPC systems, and will not have module files of any kind. While we welcome the certainty of knowing where everything is installed, users may have libraries installed in other places, and want to use them. This supports exactly the kind of outside contributions we want to encourage. So we need to continue to support and improve the CMake builds until they do everything they should to help the user build our software.

The need for standard build systems, based on CMake, is a firm requirement. It's not going away, so we need to make our CMake builds work well. ;-)

**Comment by edwardhartnett** [comment]
I propose we end this version confusion in the following way:
* Merge all changes from public-v2 into develop branch.
* When the next release is needed (public-v2.1?), changes for that release are made to develop instead of a public-v2.1 branch.
* When desired, a 2.1 release is made. We also tag it public-v2.1.
* The develop branch continues to change as it should.
* Any changes made to the public-v2.1 branch are back-ported to the develop branch as they are made.

The game then becomes to make as many of the 2.1 changes before the 2.1 release, in order to limit work on the 2.1 branch. In the ideal case, all changes have been made, no bug-fixes are needed for the well-tested code, and no modifications of the public-v2.1 branch are needed. This simplest case is the least work for everyone.

This is the way that most other projects handle this problem, and it avoids the version confusion we have here. It is also better in that it allows/forces the developers to deal with the exact same package the end-user will be dealing with. This aids in support and defect detection.

**Comment by climbfuji** [comment]
> I propose we end this version confusion in the following way:
> 
> * Merge all changes from public-v2 into develop branch.
> * When the next release is needed (public-v2.1?), changes for that release are made to develop instead of a public-v2.1 branch.
> * When desired, a 2.1 release is made. We also tag it public-v2.1.
> * The develop branch continues to change as it should.
> * Any changes made to the public-v2.1 branch are back-ported to the develop branch as they are made.
> 
> The game then becomes to make as many of the 2.1 changes before the 2.1 release, in order to limit work on the 2.1 branch. In the ideal case, all changes have been made, no bug-fixes are needed for the well-tested code, and no modifications of the public-v2.1 branch are needed. This simplest case is the least work for everyone.
> 
> This is the way that most other projects handle this problem, and it avoids the version confusion we have here. It is also better in that it allows/forces the developers to deal with the exact same package the end-user will be dealing with. This aids in support and defect detection.

We need to discuss this in a telecon with more people, writing all this up takes too long. The convention that was made for the public releases is that *one* branch release/public-vX is made to support *all* releases X.Y.Z. The next release branch to be cut from develop would thus be release/public-v3 - possibly together with the weather-s2s app release next year.

If the group comes down to using NCEPLIBs for public dissemination (which is different from what I heard in the past, namely to use hpc-stack as a step towards unifying with jedi-stack, which IS used for public releases), then that's fine and I will feed some (not all!) of release/public-v2 back to develop.

Since we support different apps with different build systems and cmake peculiarities, not everything in release/public-v2 will be applicable to release/public-v3.

Anyway, enough of the ping pong, let's get back to work!

**Comment by edwardhartnett** [comment]
OK, good, now we are getting to the heart of the confusion.

You say:

> The convention that was made for the public releases is that one branch release/public-vX is made to support all releases X.Y.Z.

Why are we attempting to use git in a unique and interesting way like this? 

OK, I think I get it. This is a consequence of distributing the code via git. In order to simplify that, they want a branch name that works everywhere for a long time. 

But that does not change the fact that the develop branch should have all the changes on the public-v2 branch.

**Comment by climbfuji** [comment]
> OK, good, now we are getting to the heart of the confusion.
> 
> You say:
> 
> > The convention that was made for the public releases is that one branch release/public-vX is made to support all releases X.Y.Z.
> 
> Why are we attempting to use git in a unique and interesting way like this?
> 
> OK, I think I get it. This is a consequence of distributing the code via git. In order to simplify that, they want a branch name that works everywhere for a long time.

This decision was made by the UFS release team leads and UFS code managers when the first public release was prepared. The same branch names are used for ufs-weather-model etc, allows us to keep one branch open for subsequent minor and bugfix releases. Branch will be closed/deleted once the next major release of that particular UFS APP is out the door and we agree that the previous major release is no longer supported. This means for utilities like NCEPLIBS that we need to keep alive multiple release branches (one for each APP), at least for now. This is all to accommodate the build system differences, workflow differences and other differences between the APPs.

As we strive for unifying these, the need to support multiple release branches will decrease.

> 
> But that does not change the fact that the develop branch should have all the changes on the public-v2 branch.

Definitely the change from lua to tcl modulefiles, yes. The NCEPLIBS developers may reject some of the others, at least that happened with release/public-v1.

**Comment by edwardhartnett** [comment]
OK, thanks for the explanations.

The NCEPLIBS developers will accommodate all necessary changes. That's what we are here to do. ;-)

**Comment by edwardhartnett** [comment]
OK, I'm a little confused about whether public-v2 changes need to be merged to develop. Do they?

Or should this issue be closed?

**Comment by climbfuji** [comment]
Per @aerorahul , no.
